### DISCUSSION TITLE
RFC NgRxTitleStrategy

### DISCUSSION THREAD

**Discussion by david-shortman** [ngrx/platform#3494] [Ideas]
Link to blog post explaining feature- https://dev.to/davidshortman/how-to-use-dynamic-titles-from-ngrx-selectors-1f0

Angular 14 introduced the ability to set a static `title` per route. In a reactive application using NgRx, developers may want the flexibility to have dynamic titles driven by the global state.

With an `NgRxTitleStrategy`, string selectors could be composed into titles:

```ts
{  // route
  title: ngrxTitle`${selectSomeDynamicTitle}`
}
```

In the app module, one could use `provideNgRxTitleStrategy` to initialize this behavior

```ts
providers: [provideNgRxTitleStrategy()]
```

This API remains small but powerfully enables titles to be based on page content/ change over time.

**Comment by david-shortman** [comment]
Here's a Stackblitz demoing the title: https://stackblitz.com/edit/angular-ivy-pbyvne?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp-routing.module.ts

Given a base route like

```ts
  {
    path: '',
    component: AppComponent,
    title: ngRxTitle`${counterFeature.selectCount} Seconds Since ${counterFeature.selectEvent}`
  },
```

Then the following title is produced:

<img width="224" alt="Screen Shot 2022-07-16 at 10 36 21" src="https://user-images.githubusercontent.com/25187726/179359356-a814a91f-4ae9-4621-883a-940323a3f6b5.png">

**Comment by david-shortman** [comment]
As an aside, the new `selectTitle` selector from @ngrx/router-store does not update when `updateTitle` is invoked from a custom title strategy:

<img width="598" alt="Screen Shot 2022-07-16 at 10 37 20" src="https://user-images.githubusercontent.com/25187726/179359456-dbc26da3-dcc6-49d9-a82b-a6b5799c81e6.png">

Seen in the [app component](https://stackblitz.com/edit/angular-ivy-pbyvne?file=src%2Fapp%2Fapp.component.ts,src%2Fapp%2Fapp-routing.module.ts):

```ts
@Component({
  selector: 'app-root',
  template: `Title selected from router store is: {{ title$ | async }}
    <br /><br />
    <button (click)="logTitle()">console.log current title</button>`,
})
export class AppComponent {
  constructor(private readonly store: Store) {}

  title$ = this.store.select(getSelectors().selectTitle);

  logTitle() {
    console.log(window.document.title);
  }
}
```

**Reply by david-shortman** [reply]
If an NgRxTitleStrategy were implemented, it could guarantee that the router store title was up-to-date by dispatching an action when the title changed.

**Comment by timdeschryver** [comment]
Hey David, smart usage of the title strategy! It's something that I didn't think of until reading your blog post. For now, I think this is very specific to one use-case, and I don't know how useful this would be for the majority. Therefore, I would suggest not including it into the NgRx repo, but this is something that we could revisit later.
